File: src/rate_of_closure/application/camera_presets.py
"""Canonical camera presets for the matched Rate clubhead viewports.

The engineering frame is always ``x=downrange, y=up, z=right``.  This
module owns orientation and scale invariants; renderers only adapt the exact
vectors to their projection API.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, replace
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from enum import StrEnum
else:
    from shared.python.compatibility import StrEnum

Vector3 = tuple[float, float, float]
MIN_ZOOM = 0.3
MAX_ZOOM = 4.0
AUTO_FIT_CLEARANCE_FRACTION = 0.16


class CameraViewId(StrEnum):
    """Stable identifiers for canonical view orientations."""

    ISOMETRIC = "camera.view.isometric"
    FACE_ON = "camera.view.face_on"
    DOWN_THE_LINE = "camera.view.down_the_line"
    OVERHEAD = "camera.view.overhead"


class CameraCommandId(StrEnum):
    """Stable identifiers for non-preset camera actions."""

    RESET_VIEW = "camera.reset_view"
    AUTO_FIT = "camera.auto_fit"


class FaceOnSide(StrEnum):
    """Physical side from which Face On observes the target line."""

    RIGHT = "right"
    LEFT = "left"


CAMERA_COMMAND_IDS: tuple[str, ...] = tuple(
    view.value for view in CameraViewId
) + tuple(command.value for command in CameraCommandId)

_ISOMETRIC_DIRECTION: Vector3 = (
    0.7071067811865476,
    -0.4082482904638631,
    -0.5773502691896258,
)
_ISOMETRIC_SCREEN_UP: Vector3 = (
    0.316227766016838,
    0.9128709291752768,
    -0.2581988897471612,
)
_VERTICAL_UP: Vector3 = (0.0, 1.0, 0.0)


@dataclass(frozen=True, slots=True)
class CameraPreset:
    """One validated toolkit-independent orientation."""

    command_id: CameraViewId
    view_direction: Vector3
    screen_up: Vector3

    def __post_init__(self) -> None:
        if not isinstance(self.command_id, CameraViewId):
            raise ValueError("unknown camera view")
        _finite_vector(self.view_direction, "view_direction")
        _finite_vector(self.screen_up, "screen_up")
        if not math.isclose(
            _norm(self.view_direction), 1.0, rel_tol=0.0, abs_tol=1e-12
        ):
            raise ValueError("view_direction must be a unit vector")
        if not math.isclose(_norm(self.screen_up), 1.0, rel_tol=0.0, abs_tol=1e-12):
            raise ValueError("screen_up must be a unit vector")
        if abs(_dot(self.view_direction, self.screen_up)) > 1e-12:
            raise ValueError("screen_up must be perpendicular to view_direction")


@dataclass(frozen=True, slots=True)
class CameraState:
    """Isolated orientation, target, scale, and explicit Face-On side."""

    preset_id: CameraViewId = CameraViewId.ISOMETRIC
    face_on_side: FaceOnSide = FaceOnSide.RIGHT
    target_m: Vector3 = (0.0, 0.0, 0.0)
    zoom: float = 1.0

    def __post_init__(self) -> None:
        if not isinstance(self.preset_id, CameraViewId):
            raise ValueError("unknown camera view")
        if not isinstance(self.face_on_side, FaceOnSide):
            raise ValueError("unknown face-on side")
        _finite_vector(self.target_m, "target_m")
        if not _finite_number(self.zoom) or not MIN_ZOOM <= self.zoom <= MAX_ZOOM:
            raise ValueError(f"zoom must be finite and within [{MIN_ZOOM}, {MAX_ZOOM}]")


def camera_preset(
    command_id: CameraViewId | str, side: FaceOnSide | str
) -> CameraPreset:
    """Return an exact canonical preset and fail closed for unknown inputs."""
    view_id = _camera_view_id(command_id)
    face_side = _face_on_side(side)
    if view_id is CameraViewId.ISOMETRIC:
        return CameraPreset(view_id, _ISOMETRIC_DIRECTION, _ISOMETRIC_SCREEN_UP)
    if view_id is CameraViewId.FACE_ON:
        direction = (0.0, 0.0, -1.0 if face_side is FaceOnSide.RIGHT else 1.0)
        return CameraPreset(view_id, direction, _VERTICAL_UP)
    if view_id is CameraViewId.DOWN_THE_LINE:
        return CameraPreset(view_id, (1.0, 0.0, 0.0), _VERTICAL_UP)
    return CameraPreset(view_id, (0.0, -1.0, 0.0), (1.0, 0.0, 0.0))


def canvas_angles(preset: CameraPreset) -> tuple[float, float]:
    """Return dependency-free canvas yaw and pitch in radians."""
    downrange, up, right = preset.view_direction
    return math.atan2(right, downrange), math.asin(max(-1.0, min(1.0, up)))


def matplotlib_angles(preset: CameraPreset) -> tuple[float, float]:
    """Return Matplotlib elevation/azimuth for display axes ``z,x,y``."""
    downrange, up, right = preset.view_direction
    if preset.command_id is CameraViewId.OVERHEAD:
        return 90.0, -90.0
    camera_display = (-right, -downrange, -up)
    return (
        math.degrees(math.asin(camera_display[2])),
        math.degrees(math.atan2(camera_display[1], camera_display[0])),
    )


def apply_camera_view(
    state: CameraState, command_id: CameraViewId | str
) -> CameraState:
    """Select a view idempotently without changing target or scale."""
    view_id = _camera_view_id(command_id)
    camera_preset(view_id, state.face_on_side)
    return replace(state, preset_id=view_id)


def set_face_on_side(state: CameraState, side: FaceOnSide | str) -> CameraState:
    """Select a deliberate lateral side without inferring handedness."""
    return replace(state, face_on_side=_face_on_side(side))


def with_camera_zoom(state: CameraState, zoom: float) -> CameraState:
    """Return ``state`` with finite zoom clamped to the supported range."""
    if not _finite_number(zoom):
        raise ValueError("zoom must be finite")
    return replace(state, zoom=max(MIN_ZOOM, min(MAX_ZOOM, zoom)))


def auto_fit_camera(
    state: CameraState,
    subject_radius_m: float,
    base_half_extent_m: float,
    clearance_fraction: float = AUTO_FIT_CLEARANCE_FRACTION,
) -> CameraState:
    """Fit a bounding sphere while changing only dimensionless zoom."""
    if not all(
        _finite_number(value) and value > 0.0
        for value in (subject_radius_m, base_half_extent_m)
    ):
        raise ValueError(
            "subject radius and base half extent must be finite and positive"
        )
    if not _finite_number(clearance_fraction) or not 0.0 <= clearance_fraction < 1.0:
        raise ValueError("clearance_fraction must be finite and within [0, 1)")
    fitted = base_half_extent_m * (1.0 - clearance_fraction) / subject_radius_m
    return with_camera_zoom(state, fitted)


def _camera_view_id(value: CameraViewId | str) -> CameraViewId:
    try:
        return value if isinstance(value, CameraViewId) else CameraViewId(value)
    except (TypeError, ValueError) as exc:
        raise ValueError(f"unknown camera view: {value!r}") from exc


def _face_on_side(value: FaceOnSide | str) -> FaceOnSide:
    try:
        return value if isinstance(value, FaceOnSide) else FaceOnSide(value)
    except (TypeError, ValueError) as exc:
        raise ValueError(f"unknown face-on side: {value!r}") from exc


def _finite_vector(vector: Vector3, name: str) -> None:
    if len(vector) != 3 or not all(_finite_number(value) for value in vector):
        raise ValueError(f"{name} must contain three finite values")


def _finite_number(value: object) -> bool:
    return (
        isinstance(value, (int, float))
        and not isinstance(value, bool)
        and math.isfinite(value)
    )


def _dot(first: Vector3, second: Vector3) -> float:
    return sum(left * right for left, right in zip(first, second, strict=True))


def _norm(vector: Vector3) -> float:
    return math.sqrt(_dot(vector, vector))


__all__ = [
    "AUTO_FIT_CLEARANCE_FRACTION",
    "CAMERA_COMMAND_IDS",
    "MAX_ZOOM",
    "MIN_ZOOM",
    "CameraCommandId",
    "CameraPreset",
    "CameraState",
    "CameraViewId",
    "FaceOnSide",
    "apply_camera_view",
    "auto_fit_camera",
    "camera_preset",
    "canvas_angles",
    "matplotlib_angles",
    "set_face_on_side",
    "with_camera_zoom",
]

File: src/rate_of_closure/web/src/model/cameraPresets.ts
/** Canonical camera presets in app frame x=downrange, y=up, z=right. */

import type { Vec3 } from "./simulation";

export const CAMERA_VIEW_IDS = [
  "camera.view.isometric",
  "camera.view.face_on",
  "camera.view.down_the_line",
  "camera.view.overhead",
] as const;
export const CAMERA_ACTION_IDS = ["camera.reset_view", "camera.auto_fit"] as const;
export const CAMERA_TRACKING_COMMAND_IDS = [
  "camera.track_clubhead", "camera.recenter",
] as const;
export const CAMERA_CONTROL_IDS = ["camera.auto_fit_fallback"] as const;
export const CAMERA_TRACKING_STATE_IDS = [
  "camera.tracking.off", "camera.tracking.active", "camera.tracking.suspended",
] as const;
export const CAMERA_PRESET_COMMAND_IDS = [
  ...CAMERA_VIEW_IDS, ...CAMERA_ACTION_IDS,
] as const;
export const CAMERA_COMMAND_IDS = [
  ...CAMERA_PRESET_COMMAND_IDS, ...CAMERA_TRACKING_COMMAND_IDS,
] as const;

export type CameraViewId = (typeof CAMERA_VIEW_IDS)[number];
export type CameraActionId = (typeof CAMERA_ACTION_IDS)[number];
export type CameraCommandId = (typeof CAMERA_COMMAND_IDS)[number];
export type CameraTrackingStateId = (typeof CAMERA_TRACKING_STATE_IDS)[number];
export type FaceOnSide = "right" | "left";

export interface CameraPreset {
  commandId: CameraViewId;
  viewDirection: Vec3;
  screenUp: Vec3;
}

export interface CameraState {
  presetId: CameraViewId;
  faceOnSide: FaceOnSide;
  targetM: Vec3;
  zoom: number;
  yawRad: number;
  pitchRad: number;
  trackingEnabled: boolean;
  trackingSuspended: boolean;
  autoFitFallbackEnabled: boolean;
}

export const MIN_CAMERA_ZOOM = 0.3;
export const MAX_CAMERA_ZOOM = 4;
export const AUTO_FIT_CLEARANCE_FRACTION = 0.16;
export const TRACKING_MAX_TARGET_STEP_M = 0.05;

const ISOMETRIC_DIRECTION: Vec3 = [
  0.7071067811865476, -0.4082482904638631, -0.5773502691896258,
];
const ISOMETRIC_SCREEN_UP: Vec3 = [
  0.316227766016838, 0.9128709291752768, -0.2581988897471612,
];
const VERTICAL_UP: Vec3 = [0, 1, 0];

export function cameraPreset(commandId: CameraViewId | string, side: FaceOnSide | string): CameraPreset {
  const viewId = parseCameraViewId(commandId);
  const faceSide = parseFaceOnSide(side);
  if (viewId === "camera.view.isometric") {
    return validatedPreset(viewId, ISOMETRIC_DIRECTION, ISOMETRIC_SCREEN_UP);
  }
  if (viewId === "camera.view.face_on") {
    return validatedPreset(viewId, [0, 0, faceSide === "right" ? -1 : 1], VERTICAL_UP);
  }
  if (viewId === "camera.view.down_the_line") {
    return validatedPreset(viewId, [1, 0, 0], VERTICAL_UP);
  }
  return validatedPreset(viewId, [0, -1, 0], [1, 0, 0]);
}

export function canvasAngles(preset: CameraPreset): { yawRad: number; pitchRad: number } {
  const [downrange, up, right] = preset.viewDirection;
  return {
    yawRad: Math.atan2(right, downrange),
    pitchRad: Math.asin(Math.max(-1, Math.min(1, up))),
  };
}

export function defaultCameraState(): CameraState {
  const preset = cameraPreset("camera.view.isometric", "right");
  return validateCameraState({
    presetId: preset.commandId,
    faceOnSide: "right",
    targetM: [0, 0, 0],
    zoom: 1,
    trackingEnabled: false,
    trackingSuspended: false,
    autoFitFallbackEnabled: false,
    ...canvasAngles(preset),
  });
}

export function applyCameraView(state: CameraState, commandId: CameraViewId | string): CameraState {
  validateCameraState(state);
  const preset = cameraPreset(commandId, state.faceOnSide);
  return { ...state, presetId: preset.commandId, ...canvasAngles(preset) };
}

export function setFaceOnSide(state: CameraState, side: FaceOnSide | string): CameraState {
  validateCameraState(state);
  const faceOnSide = parseFaceOnSide(side);
  const next = { ...state, faceOnSide };
  return state.presetId === "camera.view.face_on"
    ? applyCameraView(next, "camera.view.face_on") : next;
}

export function withCameraZoom(state: CameraState, zoom: number): CameraState {
  validateCameraState(state);
  if (!Number.isFinite(zoom)) throw new RangeError("zoom must be finite");
  return { ...state, zoom: Math.max(MIN_CAMERA_ZOOM, Math.min(MAX_CAMERA_ZOOM, zoom)) };
}

export function autoFitCamera(
  state: CameraState,
  subjectRadiusM: number,
  baseHalfExtentM: number,
  clearanceFraction = AUTO_FIT_CLEARANCE_FRACTION,
): CameraState {
  if (![subjectRadiusM, baseHalfExtentM].every((value) => Number.isFinite(value) && value > 0)) {
    throw new RangeError("subject radius and base half extent must be finite and positive");
  }
  if (!Number.isFinite(clearanceFraction) || clearanceFraction < 0 || clearanceFraction >= 1) {
    throw new RangeError("clearanceFraction must be finite and within [0, 1)");
  }
  return withCameraZoom(
    state,
    baseHalfExtentM * (1 - clearanceFraction) / subjectRadiusM,
  );
}

export function setCameraTracking(
  state: CameraState, enabled: boolean, subjectM: Vec3,
): CameraState {
  validateCameraState(state);
  if (typeof enabled !== "boolean") throw new TypeError("enabled must be a Boolean");
  validateVector(subjectM, "subjectM");
  return {
    ...state,
    targetM: enabled ? [...subjectM] : state.targetM,
    trackingEnabled: enabled,
    trackingSuspended: false,
  };
}

export function setAutoFitFallback(state: CameraState, enabled: boolean): CameraState {
  validateCameraState(state);
  if (typeof enabled !== "boolean") throw new TypeError("enabled must be a Boolean");
  return { ...state, autoFitFallbackEnabled: enabled };
}

export function applyManualCameraOverride(
  state: CameraState, targetM: Vec3 = state.targetM,
): CameraState {
  validateCameraState(state);
  validateVector(targetM, "targetM");
  return {
    ...state,
    targetM: [...targetM],
    trackingSuspended: state.trackingEnabled,
  };
}

export function recenterCamera(state: CameraState, subjectM: Vec3): CameraState {
  validateCameraState(state);
  validateVector(subjectM, "subjectM");
  return { ...state, targetM: [...subjectM], trackingSuspended: false };
}

export function updateTrackingTarget(
  state: CameraState,
  subjectM: Vec3,
  maxStepM = TRACKING_MAX_TARGET_STEP_M,
): CameraState {
  validateCameraState(state);
  validateVector(subjectM, "subjectM");
  if (!Number.isFinite(maxStepM) || maxStepM <= 0) {
    throw new RangeError("maxStepM must be finite and positive");
  }
  if (!state.trackingEnabled || state.trackingSuspended) return state;
  const delta: Vec3 = [
    subjectM[0] - state.targetM[0],
    subjectM[1] - state.targetM[1],
    subjectM[2] - state.targetM[2],
  ];
  const distance = Math.hypot(...delta);
  if (distance <= 1e-12) return state;
  const fraction = Math.min(1, maxStepM / distance);
  return {
    ...state,
    targetM: [
      state.targetM[0] + fraction * delta[0],
      state.targetM[1] + fraction * delta[1],
      state.targetM[2] + fraction * delta[2],
    ],
  };
}

export function enforceTrackingClearance(
  state: CameraState,
  subjectRadiusM: number,
  baseHalfExtentM: number,
  clearanceFraction = AUTO_FIT_CLEARANCE_FRACTION,
): CameraState {
  const fitted = autoFitCamera(
    state, subjectRadiusM, baseHalfExtentM, clearanceFraction,
  );
  return !state.autoFitFallbackEnabled || state.zoom <= fitted.zoom ? state : fitted;
}

export function trackingStateId(state: CameraState): CameraTrackingStateId {
  validateCameraState(state);
  if (!state.trackingEnabled) return "camera.tracking.off";
  return state.trackingSuspended
    ? "camera.tracking.suspended" : "camera.tracking.active";
}

function parseCameraViewId(value: string): CameraViewId {
  if (!(CAMERA_VIEW_IDS as readonly string[]).includes(value)) {
    throw new RangeError(`unknown camera view: ${JSON.stringify(value)}`);
  }
  return value as CameraViewId;
}

function parseFaceOnSide(value: string): FaceOnSide {
  if (value !== "right" && value !== "left") {
    throw new RangeError(`unknown face-on side: ${JSON.stringify(value)}`);
  }
  return value;
}

function validatedPreset(commandId: CameraViewId, viewDirection: Vec3, screenUp: Vec3): CameraPreset {
  return validateCameraPreset({ commandId, viewDirection, screenUp });
}

export function validateCameraPreset(preset: CameraPreset): CameraPreset {
  const { commandId, viewDirection, screenUp } = preset;
  parseCameraViewId(commandId);
  validateVector(viewDirection, "viewDirection");
  validateVector(screenUp, "screenUp");
  const viewNorm = Math.hypot(...viewDirection);
  const upNorm = Math.hypot(...screenUp);
  const dot = viewDirection.reduce((sum, value, axis) => sum + value * screenUp[axis], 0);
  if (Math.abs(viewNorm - 1) > 1e-12 || Math.abs(upNorm - 1) > 1e-12) {
    throw new RangeError("camera directions must be unit vectors");
  }
  if (Math.abs(dot) > 1e-12) {
    throw new RangeError("screenUp must be perpendicular to viewDirection");
  }
  return { commandId, viewDirection: [...viewDirection], screenUp: [...screenUp] };
}

export function validateCameraState(state: CameraState): CameraState {
  const presetId = parseCameraViewId(state.presetId);
  const faceOnSide = parseFaceOnSide(state.faceOnSide);
  validateVector(state.targetM, "targetM");
  if (!Number.isFinite(state.zoom)
    || state.zoom < MIN_CAMERA_ZOOM || state.zoom > MAX_CAMERA_ZOOM) {
    throw new RangeError(
      `zoom must be finite and within [${MIN_CAMERA_ZOOM}, ${MAX_CAMERA_ZOOM}]`,
    );
  }
  if (!Number.isFinite(state.yawRad) || !Number.isFinite(state.pitchRad)) {
    throw new RangeError("camera angles must be finite");
  }
  for (const [value, name] of [
    [state.trackingEnabled, "trackingEnabled"],
    [state.trackingSuspended, "trackingSuspended"],
    [state.autoFitFallbackEnabled, "autoFitFallbackEnabled"],
  ] as const) {
    if (typeof value !== "boolean") throw new TypeError(`${name} must be a Boolean`);
  }
  if (state.trackingSuspended && !state.trackingEnabled) {
    throw new RangeError("tracking cannot be suspended while disabled");
  }
  return {
    ...state,
    presetId,
    faceOnSide,
    targetM: [...state.targetM],
  };
}

function validateVector(vector: Vec3, name: string): void {
  if (vector.length !== 3 || !vector.every(Number.isFinite)) {
    throw new RangeError(`${name} must contain three finite values`);
  }
}
